File: Test/cs/Serialization.cs
using System;
using SpacyDotNet;

namespace Test
{
    static class Serialization
    {
        public static void PrintDoc(Doc adoc)
        {
            foreach (Token word in adoc.Tokens)
            {
                var lexeme = adoc.Vocab[word.Text];
                Console.WriteLine($@"{lexeme.Text} {lexeme.Orth} {lexeme.Shape} {lexeme.Prefix} {lexeme.Suffix} {lexeme.IsAlpha} {lexeme.IsDigit} {lexeme.IsTitle} {lexeme.Lang}");
            }
        }

        public static void Run()
        {
            var spacy = new Spacy();

            var text = "I love coffee";

            // Load base document
            var nlp = spacy.Load("en_core_web_sm");
            var docBase = nlp.GetDocument(text);
            Console.WriteLine("");
            PrintDoc(docBase);

            // Serialize document to disk and bytes
            docBase.ToDisk("doc.out");
            var docBaseBytes = docBase.ToBytes();

            // Serialize using DocBin
            var docBinBase = new DocBin(attrs: new string[] { "ENT_IOB", "POS", "HEAD", "DEP", "ENT_TYPE" }, storeUserData: true);
            docBinBase.Add(docBase);
            var docBinBaseBytes = docBinBase.ToBytes();

            // Restore document from disk
            var doc = new Doc(new Vocab());
            doc.FromDisk("doc.out");
            Console.WriteLine("");
            PrintDoc(doc);

            // Restore document from bytes
            doc = new Doc(new Vocab());
            doc.FromBytes(docBaseBytes);
            Console.WriteLine("");
            PrintDoc(doc);

            // Restore using DocBin
            var docBin = new DocBin();
            docBin.FromBytes(docBinBaseBytes);
            var docs = docBin.GetDocs(nlp.Vocab);
            Console.WriteLine("");
            PrintDoc(docs[0]);
        }
    }
}

File: Test/py/serialization.py
import spacy

from spacy.tokens import DocBin
from spacy.tokens import Doc
from spacy.vocab import Vocab

def print_doc(adoc):
    for word in adoc:
        lexeme = adoc.vocab[word.text]
        print(lexeme.text, lexeme.orth, lexeme.shape_, lexeme.prefix_, lexeme.suffix_,
              lexeme.is_alpha, lexeme.is_digit, lexeme.is_title, lexeme.lang_)

text = "I love coffee"

# Load base document
nlp = spacy.load("en_core_web_sm")
doc_base = nlp(text)
print("")
print_doc(doc_base)

# Serialize document to disk and bytes
doc_base.to_disk("doc.spacy")
doc_base_bytes = doc_base.to_bytes()

# Serialize using DocBin
docbin_base = DocBin(attrs=["ENT_IOB", "POS", "HEAD", "DEP", "ENT_TYPE"], store_user_data=True)
docbin_base.add(doc_base)
docbin_base_bytes = docbin_base.to_bytes()

# Restore document from disk
doc = Doc(Vocab())
doc.from_disk("doc.spacy")
print("")
print_doc(doc)

# Restore document from bytes
doc = Doc(Vocab())
doc.from_bytes(doc_base_bytes)
print("")
print_doc(doc)

# Restore using DocBin
docbin = DocBin().from_bytes(docbin_base_bytes)
docs = list(docbin.get_docs(nlp.vocab))
print("")
print_doc(docs[0])